src/Connection.php

```diff
@@ -1283,16 +1284,30 @@ public function transactional(Closure $func)
 
         try {
             $res = $func($this);
-            $this->commit();
 
             $successful = true;
-
-            return $res;
         } finally {
             if (! $successful) {
                 $this->rollBack();
             }
         }
+
+        $shouldRollback = true;
+        try {
+            $this->commit();
+
+            $shouldRollback = false;
+        } catch (TheDriverException $t) {
+            $shouldRollback = false;
+
+            throw $t;
+        } finally {
+            if ($shouldRollback) {
+                $this->rollBack();
+            }
+        }
```

Comment:
Why not do something similar to what I did in https://github.com/doctrine/orm/pull/11646/commits/b6137c89110952664cdc0b2e4df7f98716ff3f93 and use `isTransactionActive()`?

Reply:
I forgot about that API, should work.

Reply:
What should happen in this scenario?

```php
$connection->beginTransaction();
$connection->transactional(some function call here);
$connection->commit();
```

I'm not familiar with nested transactions, but if it's possible to nest them then this will return true because of the outer transaction, although the outer transaction will have been property closed. Also, why not reuse the previous `try` / `finally` block just like I did in ORM?

Reply:
Nested transactions should be covered by 4 tests.

I merged try blocks, it makes sense and is nice with the `isTransactionActive` API.

Reply:
Now the code looks like it is going to wrongly revert the outer transaction after successfully committing the inner one, just because the outer one is active (as it should)

Reply:
I'm surprised the test suite is green, maybe I'm missing something.

Reply:
Haha that's funny. I'll add a test case to verify it but I think you're right.

Reply:
Umm I'm confused now. Will take a break.

Reply:
I can add a check of nesting level but the green test suite somehow implies it won't happen at all anyway.

Reply:
It indeed tried to rollback the transaction. I have reverted to a previous version and added a test to cover the case.

Reply:
I am confused now. Are you saying that the bug exists, and that you reproduced it? I don't see any check on the transaction nesting level, which does sound exactly like what would be needed here, so why is the tests suite mostly green (it's red on Oracle, do you know why?)

Reply:
> it's red on Oracle, do you know why

oracle and db2 require `select...` to have also `from` clause. Fixed.

> Are you saying that the bug exists, and that you reproduced it

Yes, the bug would have been introduced with 88879b3ad5cabff479e1e85a2975091887f2744e. I covered it with `testNestedTransactionWalkthrough()` so we know it passes.

When doing `commit()`, we need to rollback the transaction only if something failed outside the database https://github.com/doctrine/dbal/pull/6545#discussion_r1797695417.

Since we need to have `rollback()` in `finally`, we need to use boolean flag to know whether to trigger it.  
We can't use `isTransactionActive()` since it only checks the nesting level. `isTransactionActive()` can actually be renamed to something `isTransactionNested()`.